engine/src/witness/generic_elections.rs

```diff
@@ -0,0 +1,233 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use cf_utilities::task_scope::{self, Scope};
+use futures::FutureExt;
+use pallet_cf_elections::{
+	electoral_system::ElectoralSystemTypes,
+	electoral_systems::oracle_price::{
+		chainlink::{ChainlinkAssetpair, ChainlinkPrice},
+		primitives::UnixTime,
+		state_machine::{should_vote_for_asset, ExternalChainStateVote, ExternalPriceChain},
+	},
+	VoteOf,
+};
+use sp_core::U256;
+use state_chain_runtime::chainflip::generic_elections::*;
+
+use crate::{
+	elections::voter_api::{CompositeVoter, VoterApi},
+	evm::{
+		retry_rpc::{address_checker::AddressCheckerRetryRpcApi, EvmRetryRpcClient},
+		rpc::{address_checker::PriceFeedData as EthPriceFeedData, EvmRpcSigningClient},
+	},
+	sol::retry_rpc::SolRetryRpcClient,
+	state_chain_observer::client::{
+		chain_api::ChainApi, electoral_api::ElectoralApi,
+		extrinsic_api::signed::SignedExtrinsicApi, storage_api::StorageApi,
+	},
+	witness::sol::oracle_witnessing::get_price_feeds,
+};
+use anyhow::Result;
+use sol_prim::program_instructions::PriceFeedData as SolPriceFeedData;
+
+/// IMPORTANT: These strings have to match with the price feed "description" as returned by
+/// chainlink.
+pub fn asset_pair_from_description(description: String) -> Option<ChainlinkAssetpair> {
+	use ChainlinkAssetpair::*;
+	match description.as_str() {
+		"BTC / USD" => Some(BtcUsd),
+		"ETH / USD" => Some(EthUsd),
+		"SOL / USD" => Some(SolUsd),
+		"USDT / USD" => Some(UsdtUsd),
+		"USDC / USD" => Some(UsdcUsd),
+		_ => None,
+	}
+}
+
+#[derive(Clone)]
+struct OraclePriceVoter {
+	sol_client: SolRetryRpcClient,
+	eth_client: EvmRetryRpcClient<EvmRpcSigningClient>,
+}
+
+#[derive(Debug, Clone)]
+struct PriceData {
+	pub description: String,
+	pub answer: i128,
+	pub decimals: u8,
+	pub timestamp: UnixTime,
+}
+
+impl From<SolPriceFeedData> for PriceData {
+	fn from(value: SolPriceFeedData) -> Self {
+		let SolPriceFeedData { round_id: _, slot: _, timestamp, answer, decimals, description } =
+			value;
+		Self { description, answer, decimals, timestamp: UnixTime { seconds: timestamp as u64 } }
+	}
+}
+
+impl From<EthPriceFeedData> for PriceData {
+	fn from(value: EthPriceFeedData) -> Self {
+		let EthPriceFeedData {
+			round_id: _,
+			answer,
+			started_at: _,
+			updated_at,
+			answered_in_round: _,
+			decimals,
+			description,
+		} = value;
+		Self {
+			description,
+			answer: answer.try_into().unwrap(),
+			decimals,
+			timestamp: UnixTime { seconds: updated_at.try_into().unwrap() },
```

Comment:
I can't find the PriceFeedData type... If these unwraps are safe, would be good to explain how/why in an `expect`.

Reply:
Good catch, appearently the `answer` is an `I256`.